Month3 Project Week 2.1Linked List

camodesigns opened the pull request:
Operators [] created
Head() Tail () Created
Created
IsEmpty()
GetSize()
AddHead()
AddTail()
insert()
Remove()
Clear()
 FindByPredicate()
 FilterByPredicate()
 RemoveAllByPredicate() 
begin()
end()
 operator++, operator* y operator !=

camodesigns committed:
Exercise 1

Class TSSList created
struct FNode created
TElement and pointers created
T operator[]() create

camodesigns committed:
Exercise2-4

Created
operator[]()
GetHead()
getTail()
IsEmpty()
GetSize()
AddHead()
AddTail()
Insert()
Remove()
Clear()
const version

camodesigns committed:
Exercise5

Created
CopyConstructor

camodesigns committed:
Part 2 LinkedList Project

 Methods createds
Updated Methods and   corrections added

Sergio6040 commented on Month3/LinkedList/LinkedList/TSLList.h:
```diff
@@ -0,0 +1,334 @@
+	TSLList()
+	{
+
+	}
```
this works, but you can use the constructor to define the default values.

camodesigns replied:
I define mutiples constructors whit diferent parameters, Is not necesary pass parameteres in all the constructors

Sergio6040 commented on Month3/LinkedList/LinkedList/TSLList.h:
```diff
@@ -0,0 +1,334 @@
+		}
+	}
+
+	TSLList& operator=(const TSLList<T>& OtherList)
+	{
+		Clear();
+		FNode* Current = OtherList.Head;
+		for (int CurrentIndex = 0; CurrentIndex < OtherList.Size; CurrentIndex++)
+		{
+			AddTail(Current->Element);
+			Current = Current->Next;
+		}
+		return *this;
+	}
+
```
well done, but i think there si no need of the return, this could be void.

camodesigns replied:
jaja please check your writing, si(is) no need, and  in my case I believe in this case is better return

Sergio6040 commented on Month3/LinkedList/LinkedList/TSLList.h:
```diff
@@ -0,0 +1,334 @@
+
+	void Insert(const T& NewElement, const int Index )
+	{
+		if (Index < 0 || Index > Size) 
+		{
+			return;
+		}
```
Great, but I think it's easier to understand if you do the correct case :)
what I mean is:
```
if(Index > 0 && Index < Size)
{...
 ```

Sergio6040 replied:
David, said in my code review that this way it's easier to understand xd, now is in your hands xd

camodesigns replied:
in my code I validate first and after this I make dthe function, in this context is understandable

Sergio6040 commented on Month3/LinkedList/LinkedList/TSLList.h:
```diff
@@ -0,0 +1,334 @@
+
+	void Remove(const int Index) 
+	{
+		if (Index >= Size || Index < 0) 
+		{
+			return;
+		}
```
same here about the correct case

camodesigns replied:
Again the context

Sergio6040 commented on Month3/LinkedList/LinkedList/TSLList.h:
```diff
@@ -0,0 +1,334 @@
+
+	T& operator[](const int Index) 
+	{
+		FNode* Current = Head;
+		for (int CurrentIndex = 0; CurrentIndex < Index; CurrentIndex++) 
+		{
+			Current = Current->Next;
+		}
+		
```
here you can create a GetNode function, to not repeat code :)

Sergio6040 replied:
if you wanna xd anyway this is fine

camodesigns replied:
Is other way, but I prefer this way

Sergio6040 commented on Month3/LinkedList/LinkedList/TSLList.h:
```diff
@@ -0,0 +1,334 @@
+
+	template<typename Pred>
+
+	T& FindByPredicate(const Pred& Predicate) 
```
the return type should be *, shouldn't it?

Sergio6040 commented on Month3/LinkedList/LinkedList/TSLList.h:
```diff
@@ -0,0 +1,334 @@
+		FNode* Current = Head;
+		for (int CurrentIndex = 0; CurrentIndex < Size; CurrentIndex++) 
+		{
+			T Element = Current->Element;
```
you are not using this T Element.

Sergio6040 commented on Month3/LinkedList/LinkedList/TSLList.h:
```diff
@@ -0,0 +1,334 @@
+		{
+			T Element = Current->Element;
+			Current = Current->Next;
+			if (Predicate(Current->Element))
```
you are not passing the head to the Predicate function the `Current = Current->Next` should be before the if statment

Sergio6040 commented on Month3/LinkedList/LinkedList/TSLList.h:
```diff
@@ -0,0 +1,334 @@
+			if (Predicate(Current->Element))
+			{
+				Remove(CurrentIndex);
+				CurrentIndex--;
```
Interesting one, I just realize that if this works, I have an error in my code xd

camodesigns replied:
lol

Sergio6040 commented on Month3/LinkedList/LinkedList/TSLList.h:
```diff
@@ -0,0 +1,334 @@
+			return Current != CurrentOther.Current;
+		}
+
+		const bool operator!=(FIterator& CurrentOther) const
```
the parameter should be const

Sergio6040 commented on Month3/LinkedList/LinkedList/TSLList.h:
```diff
@@ -0,0 +1,334 @@
+			return *this;
+		}
+
+		T& operator*()
+		{
+			return Current->Element;
+		}
+
+		const T& operator*() const
+		{
+			return Current->Element;
+		}
```
Good job 👍

Sergio6040 commented on Month3/LinkedList/LinkedList/TSLList.h:
```diff
@@ -0,0 +1,334 @@
+#pragma once
+# include <iostream>
+#include <string>
+#include <functional>
```
Are you using these libraries?

camodesigns replied:
good comment

Sergio6040 commented on Month3/LinkedList/LinkedList/TSLList.h:
```diff
@@ -0,0 +1,334 @@
+	{
+		return Size == 0;
+	}
+
```
remember to use separators between functions, coding standards 😉

camodesigns replied:
ok

Sergio6040 commented on Month3/LinkedList/LinkedList/TSLList.cpp:
```diff
@@ -0,0 +1,178 @@
+	std::cout << "Tail is:\t" << List.GetTail() << (List.GetTail() == CorrectTail ? " CORRECT" : " INCORRECT") << std::endl;
+}
+
+int main()
```
this file should be called main or source, anything but TSLList is confusing

camodesigns replied:
itself is the application of using SLList, so in that context it should not be confusing

Sergio6040 reviewed (commented):
Great work in general, just some little comments 👌👍

camodesigns committed:
Complete project

Created:
class TStack
Created:
Top()
const Top()
Push()
IsEmpty()
Pop()
class TQueue
Created:
IsEmpty()
Front()
cosnt Front()
Enqueue()
Dequeue();
class TDLList()
this contain all the functions and methods

DvdHalliday commented on Month3/LinkedList/LinkedList/TQueue.h:
```diff
@@ -0,0 +1,46 @@
+#include "TSLList.h"
+
+template<typename T>
+
```
Unnecessary blank line

DvdHalliday commented on Month3/LinkedList/LinkedList/TQueue.h:
```diff
@@ -0,0 +1,46 @@
+public:
+	TQueue()
+	{
+	
```
Unnecessary blank line

DvdHalliday commented on Month3/LinkedList/LinkedList/TQueue.h:
```diff
@@ -0,0 +1,46 @@
+	{
+	
+	}
+	~TQueue()
+	{
+		List.Clear();
+	}
```
This is not incorrect, although you're not actually helping the system since it will automatically destroy the list calling it's defined destructor once it needs to do it

DvdHalliday commented on Month3/LinkedList/LinkedList/TQueue.h:
```diff
@@ -0,0 +1,46 @@
+		List.Clear();
+	}
+
+	bool IsEmpty()
```
const method

simonkytr replied:
maybe this function should be const

DvdHalliday commented on Month3/LinkedList/LinkedList/TStack.h:
```diff
@@ -0,0 +1,45 @@
+public:
+	TStack()
+	{
+	
```
Unnecessary blank line

camodesigns replied:
updated

DvdHalliday commented on Month3/LinkedList/LinkedList/TStack.h:
```diff
@@ -0,0 +1,45 @@
+#pragma once
+#include "TSLList.h"
+template <typename T>
+
```
Unnecessary blank line

camodesigns replied:
I want make undesrtandable

DvdHalliday replied:
Then it would be best to separate with aan empty line the includes to the template, but template usually goes along without any space with whatever structure it is a template of

DvdHalliday commented on Month3/LinkedList/LinkedList/TStack.h:
```diff
@@ -0,0 +1,45 @@
+	
+	}
+
+	int GetSize()
```
const method

DvdHalliday commented on Month3/LinkedList/LinkedList/TStack.h:
```diff
@@ -0,0 +1,45 @@
+	{
+		return List.GetSize;
+	}
+	bool IsEmpty()
```
const method

DvdHalliday commented on Month3/LinkedList/LinkedList/TStack.h:
```diff
@@ -0,0 +1,45 @@
+		return List.IsEmpty();
+	}
+
+	void Push( T& NewElement)
```
Unnecessary space befor T&

DvdHalliday replied:
also, it should be a const parameter

camodesigns replied:
updated

DvdHalliday commented on Month3/LinkedList/LinkedList/TStack.h:
```diff
@@ -0,0 +1,45 @@
+	{
+		List.RemoveHead
+	}
+
```
Unnecessary blank line

DvdHalliday commented on Month3/LinkedList/LinkedList/TDLList.h:
```diff
@@ -0,0 +1,404 @@
+#include <iostream>
+
+template<typename T>
+
```
Unnecessary line

DvdHalliday commented on Month3/LinkedList/LinkedList/TDLList.h:
```diff
@@ -0,0 +1,404 @@
+		FNode* Next = nullptr;
+		FNode* Prev = nullptr;
+	};
+	FNode* Tail = nullptr;
```
You could add some blank line here, this would actually help with readability

DvdHalliday commented on Month3/LinkedList/LinkedList/TDLList.h:
```diff
@@ -0,0 +1,404 @@
+public:
+	TDLList()
+	{
+
```
Unnecessary line

DvdHalliday commented on Month3/LinkedList/LinkedList/TDLList.h:
```diff
@@ -0,0 +1,404 @@
+	{
+
+	}
+	TDLList(T& NewElement)
```
Const parameter

DvdHalliday replied:
Did you ever use this constructor?

DvdHalliday commented on Month3/LinkedList/LinkedList/TDLList.h:
```diff
@@ -0,0 +1,404 @@
+	T& operator[](const int Index)
+	{
+		FNode* Current = Head;
+		if (Index < 0 || Index>Size)
+		{
+			return;
+		}
```
You actually cannot do this. We're not supposed to handle boundness of indexes on this operator. The real issue is that you're not returning a T element, you're just returning, and that's going to cause some trouble, you should always return something when there is a return type different to void set. Same thing applies for the const version

DvdHalliday commented on Month3/LinkedList/LinkedList/TDLList.h:
```diff
@@ -0,0 +1,404 @@
+		}
+		if (Index < (Size / 2))
+		{
+
```
Unnecessary line

DvdHalliday commented on Month3/LinkedList/LinkedList/TDLList.h:
```diff
@@ -0,0 +1,404 @@
+			}
+			return Current->Next;
+		}
+		else if (Index > (Size / 2))
```
Why would you need that other if? when the index isn't less than Size / 2 it's going to be greater than or equal than that value. Also, you're going to get an error when you actually call for the Index Size /2 only because you added this unnecessary condition

DvdHalliday commented on Month3/LinkedList/LinkedList/TDLList.h:
```diff
@@ -0,0 +1,404 @@
+	{
+		T Element;
+		FNode* Next = nullptr;
+		FNode* Prev = nullptr;
```
I'd encourage you not to shorten the key word of a variable name, Previous is less prone to confusions

simonkytr replied:
you shouldn't shorten the names

DvdHalliday commented on Month3/LinkedList/LinkedList/TDLList.h:
```diff
@@ -0,0 +1,404 @@
+		return Tail->Element;
+	}
+
+	bool IsEmpty()
```
const method

DvdHalliday commented on Month3/LinkedList/LinkedList/TDLList.h:
```diff
@@ -0,0 +1,404 @@
+		return Size == 0;
+	}
+
+	const int GetSize() const
```
return types shouldn't be const unless you're returning a reference or a pointer

DvdHalliday commented on Month3/LinkedList/LinkedList/TDLList.h:
```diff
@@ -0,0 +1,404 @@
+
+	void AddHead(const T& NewElement)
+	{
+		if (Size == 0)
```
Size == 0 is the same statement you use for IsEmpty, maybe consider using it

DvdHalliday commented on Month3/LinkedList/LinkedList/TDLList.h:
```diff
@@ -0,0 +1,404 @@
+	{
+		if (Size == 0)
+		{
+			FNode* NewNode = new FNode;
+			NewNode->Element = NewElement;
+
+			Head = NewNode;
+			Tail = NewNode;
+
+			++Size;
+		}
+		else
+		{
+			FNode* NewNode = new FNode;
+			NewNode->Element = NewElement;
+
+			Head->Prev = NewNode;
+			NewNode->Next = Head;
+			Head = NewNode;
+			++Size;
```
You're repeating lines that are exactly the same, you can do that before and or after the if-else statement to avoid code repetition

simonkytr replied:
you can put out from the statements the lines `Head = newNode`  and `Size++;`

DvdHalliday commented on Month3/LinkedList/LinkedList/TDLList.h:
```diff
@@ -0,0 +1,404 @@
+	{
+		if (Tail) {
+
+			/*FNode* NewNode = new FNode;
+			NewNode->Element = NewElement;
+			NewNode->Next = nullptr;
+			Tail->Next = NewNode;
+			NewNode->Prev = Tail;
+			Tail = NewNode;
+			Size++;*/
```
Avoid leaving unnecessary comment on the code you add to your pull requests

DvdHalliday commented on Month3/LinkedList/LinkedList/TDLList.h:
```diff
@@ -0,0 +1,404 @@
+			Tail = NewNode;
+			Size++;*/
+
+			Tail->Next = new FNode;
+			Tail->Next->Element = NewElement;
+			Tail->Next->Prev = Tail;
+			Tail = Tail->Next;
+			Size++;
```
Since you're gonna be using Tail->Next as an element you could consider naming it as a local variable just to help with readability

DvdHalliday replied:
You could also add blank lines for the same reason

DvdHalliday commented on Month3/LinkedList/LinkedList/TDLList.h:
```diff
@@ -0,0 +1,404 @@
+			return;
+		}
+
+		else if (Index == 0)
```
you can simply take that else from there and it will work the same but it makes it easier to read

DvdHalliday commented on Month3/LinkedList/LinkedList/TDLList.h:
```diff
@@ -0,0 +1,404 @@
+			AddHead(NewElement);
+		}
+
+		else if (Index == Size)
+		{
+			AddTail(NewElement);
+		}
```
I'm not sure this needs to be another condition on its own, it could be inside the next else

DvdHalliday commented on Month3/LinkedList/LinkedList/TDLList.h:
```diff
@@ -0,0 +1,404 @@
+			NewNode->Element = NewElement;
+			FNode* Current = Head;
+
+			for (int Currentindex = 0; Currentindex < Index; ++Currentindex)
```
the I in Currentindex should be capitalised

DvdHalliday replied:
Woah I noticed the use of the  pre-increment operator

DvdHalliday commented on Month3/LinkedList/LinkedList/TDLList.h:
```diff
@@ -0,0 +1,404 @@
+			}
+			NewNode->Next = Current;
+			NewNode->Prev = Current->Prev;
+			Current->Prev->Next = NewNode;
```
When you start using the -> operator twice  on the same pointer it's likely you could've done that differently and better for readability

DvdHalliday commented on Month3/LinkedList/LinkedList/TDLList.h:
```diff
@@ -0,0 +1,404 @@
+			Current->Prev = NewNode;
+			Size++;
+		}
+
+
+
+	}
```
That's one too many blank lines

DvdHalliday commented on Month3/LinkedList/LinkedList/TDLList.h:
```diff
@@ -0,0 +1,404 @@
+
+
+	}
+
+
+
```
Same here

simonkytr replied:
you can use the separator 
`//-----------------------------------------`

DvdHalliday commented on Month3/LinkedList/LinkedList/TDLList.h:
```diff
@@ -0,0 +1,404 @@
+
+	void RemoveTail()
+	{
+		FNode* Current = Tail;
```
Now this variable is useless, you could simply do delete Tail->Next after line 234

simonkytr commented on Month3/LinkedList/LinkedList/TSLList.h:
```diff
@@ -0,0 +1,349 @@
+public:
+	TSLList()
+	{
+
```
Empty space unnecessary

DvdHalliday commented on Month3/LinkedList/LinkedList/TDLList.h:
```diff
@@ -0,0 +1,404 @@
+	void RemoveTail()
+	{
+		FNode* Current = Tail;
+		Tail = Tail->Prev;
+		delete Current;
+		--Size;
```
Also, you're never setting Tail->Next to nullptr and that may cause trouble

DvdHalliday commented on Month3/LinkedList/LinkedList/TDLList.h:
```diff
@@ -0,0 +1,404 @@
+
+	void Remove(const int Index)
+	{
+		if (Index >= Size || Index < 0 || Size == 0)
```
The size == 0 condition isn't really helping at all , if size is 0 then the other 2 conditions shall already make the statement true

DvdHalliday commented on Month3/LinkedList/LinkedList/TDLList.h:
```diff
@@ -0,0 +1,404 @@
+			return;
+		}
+
+		else if (Size == 1)
```
No need for an else statement, if it reached thus far it means it hasn't returned

DvdHalliday commented on Month3/LinkedList/LinkedList/TDLList.h:
```diff
@@ -0,0 +1,404 @@
+
+		else if (Size == 1)
+		{
+			Clear();
```
Nicely used

DvdHalliday commented on Month3/LinkedList/LinkedList/TDLList.h:
```diff
@@ -0,0 +1,404 @@
+		else if (Index == (Size - 1))
+		{
+			RemoveTail();
+
```
Unnecessary line

DvdHalliday commented on Month3/LinkedList/LinkedList/TDLList.h:
```diff
@@ -0,0 +1,404 @@
+				Current = Current->Next;
+			}
+
+			FNode* After = Current->Next;
```
After isn't an adequate name for it, it has to do with time.

simonkytr replied:
maybe you can use `NextNode` [?]

DvdHalliday commented on Month3/LinkedList/LinkedList/TDLList.h:
```diff
@@ -0,0 +1,404 @@
+			}
+
+			FNode* After = Current->Next;
+			Current->Next = After->Next;
```
This is confusing, you're actually calling current the previous node you want to delete

DvdHalliday replied:
So in fact you mean to delete after

DvdHalliday replied:
The current node you want to delete should carry that name, if anything

DvdHalliday commented on Month3/LinkedList/LinkedList/TDLList.h:
```diff
@@ -0,0 +1,404 @@
+		}
+
+	}
+
+
+
```
Unnecessary lines

DvdHalliday commented on Month3/LinkedList/LinkedList/TDLList.h:
```diff
@@ -0,0 +1,404 @@
+	{
+		for (int CurrentIndex = 0; CurrentIndex < Size; ++CurrentIndex)
+		{
+			FNode* Current = Head;
+			Head = Head->Next;
+			delete Current;
```
Hmm you're not setting the new Head previous element to nullptr, but i guess it doesn't hurt anybody since they're all gonna get deleted straight away

DvdHalliday commented on Month3/LinkedList/LinkedList/TDLList.h:
```diff
@@ -0,0 +1,404 @@
+	}
+
+	template<typename Pred>
+
```
Unnecessary line

DvdHalliday commented on Month3/LinkedList/LinkedList/TDLList.h:
```diff
@@ -0,0 +1,404 @@
+			}
+			Current = Current->Next;
+		}
+		return nullptr;
```
nullptr isn't necessarily the same type T is, so you cannot return it if you want to get a T reference

DvdHalliday commented on Month3/LinkedList/LinkedList/TDLList.h:
```diff
@@ -0,0 +1,404 @@
+		{
+			if (Predicate(Current->Element))
+			{
+				return *Current;
```
*Current is dereferencing a FNode pointer, so you will get a FNode, and that is not your return type

simonkytr replied:
yo have to return current->element

DvdHalliday commented on Month3/LinkedList/LinkedList/TDLList.h:
```diff
@@ -0,0 +1,404 @@
+		FNode* Current = Head;
+		for (int CurrentIndex = 0; CurrentIndex < Size; CurrentIndex++)
+		{
+			T Element = Current->Element;
```
You never use this variable

DvdHalliday commented on Month3/LinkedList/LinkedList/TDLList.h:
```diff
@@ -0,0 +1,404 @@
+			}
+		}
+	}
+
+
+
+
+
```
Unnecessary lines

DvdHalliday commented on Month3/LinkedList/LinkedList/TDLList.h:
```diff
@@ -0,0 +1,404 @@
+	private:
+		FNode* Current = nullptr;
+	public:
+		FIterator(FNode* Currently)
```
Currently isn't an adequate name for this

DvdHalliday commented on Month3/LinkedList/LinkedList/TDLList.h:
```diff
@@ -0,0 +1,404 @@
+			return Current->Element;
+		}
+
+		bool operator!=(FIterator CurrentOther)
```
An Iterator isn't something current, there is however an attribute of it that is Current, so the name of the other Iterator shouldn't be CurrentOther

DvdHalliday commented on Month3/LinkedList/LinkedList/TDLList.h:
```diff
@@ -0,0 +1,404 @@
+		{
+			return Current != CurrentOther.Current;
+		}
+
+		const bool operator!=(FIterator& CurrentOther) const
+		{
+			return Current != CurrentOther.Current;
+		}
```
There is no need for this version of the operator

DvdHalliday commented on Month3/LinkedList/LinkedList/TDLList.h:
```diff
@@ -0,0 +1,404 @@
+			return Current != CurrentOther.Current;
+		}
+	};
+
+
+
```
Unnecessary lines

DvdHalliday commented on Month3/LinkedList/LinkedList/TDLList.h:
```diff
@@ -0,0 +1,404 @@
+	{
+		return FIterator(Head);
+	}
+
+
+
```
Unnecessary lines

DvdHalliday reviewed (changes requested):
Nicely done.

Don't forget to add separators between functions and avoid adding extra blank lines. Overall may be just try to keep your code clean, that is what the standards are for and it will help you and those who read your code a ton. Cheers

DvdHalliday commented on Month3/LinkedList/LinkedList/TQueue.h:
```diff
@@ -0,0 +1,46 @@
+		List.GetHead();
+	}
+
+	void Enqueue()
+	{
+		List.AddTail();
+	}
```
What are you adding here?

simonkytr commented on Month3/LinkedList/LinkedList/TStack.h:
```diff
@@ -0,0 +1,45 @@
+	{
+		List.AddHead(NewElement);
+	}
+
```
extra empty spaces here!

simonkytr commented on Month3/LinkedList/LinkedList/TDLList.h:
```diff
@@ -0,0 +1,404 @@
+
+	~TDLList()
+	{
+		Clear();
```
you should put the destructor closer to the constructor

simonkytr commented on Month3/LinkedList/LinkedList/TDLList.h:
```diff
@@ -0,0 +1,404 @@
+	void ForEach(const Pred& Predicate)
+	{
+		FNode* Current = Head;
+		for (int CurrentIndex = 0; CurrentIndex < Size; CurrentIndex++)
```
you can use the for like comparison
for(T& Type : *this)

simonkytr commented on Month3/LinkedList/LinkedList/TDLList.h:
```diff
@@ -0,0 +1,404 @@
+
+
+
+	class FIterator
```
you should put this class at the beginning to sort your class

simonkytr reviewed (commented):
I Hope to complete David's Corrections, keep in mind you have to watch your sort class, and this is a good moment to start implementing the separators

camodesigns committed:
updates
